homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,370 @@
+    descriptions = yield from hass.loop.run_in_executor(
+        None, load_yaml_config_file,
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    def async_setup_platform(p_type, p_config, disc_info=None):
+        """Setup a tts platform."""
+        platform = yield from async_prepare_setup_platform(
+            hass, config, DOMAIN, p_type)
+        if platform is None:
+            return
+
+        try:
+            if hasattr(platform, 'async_get_engine'):
+                provider = yield from platform.async_get_engine(
+                    hass, p_config)
+            else:
+                provider = yield from hass.loop.run_in_executor(
+                    None, platform.get_engine, hass, p_config)
+
+            if provider is None:
+                _LOGGER.error('Error setting up platform %s', p_type)
+                return
+
+            yield from tts.async_register_engine(p_type, provider, p_config)
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception('Error setting up platform %s', p_type)
+            return
+
+        @asyncio.coroutine
+        def async_say_handle(service):
+            """Service handle for say."""
+            entity_ids = service.data.get(ATTR_ENTITY_ID)
+            message = service.data.get(ATTR_MESSAGE)
+            cache = service.data.get(ATTR_CACHE)
+
+            try:
+                url = yield from tts.async_get_url(
+                    p_type, message, cache=cache)
+            except HomeAssistantError as err:
+                _LOGGER.error("Error on init tts: %s", err)
+                return
+
+            data = {
+                ATTR_MEDIA_CONTENT_ID: url,
+                ATTR_MEDIA_CONTENT_TYPE: MEDIA_TYPE_MUSIC,
+            }
+
+            if entity_ids:
+                data[ATTR_ENTITY_ID] = entity_ids
+
+            yield from hass.services.async_call(
+                DOMAIN_MP, SERVICE_PLAY_MEDIA, data, blocking=True)
+
+        hass.services.async_register(
+            DOMAIN, "{}_{}".format(p_type, SERVICE_SAY), async_say_handle,
+            descriptions.get(SERVICE_SAY))
+
+    setup_tasks = [async_setup_platform(p_type, p_config) for p_type, p_config
+                   in config_per_platform(config, DOMAIN)]
+
+    if setup_tasks:
+        yield from asyncio.wait(setup_tasks, loop=hass.loop)
+
+    return True
+
+
+class SpeechManager(object):
+    """Representation of a speech store."""
+
+    def __init__(self, hass):
+        """Initialize a speech store."""
+        self.hass = hass
+        self.file_cache = {}
+        self.mem_cache = {}
+        self.providers = {}
+        self._cache_dirs = {}
+
+    @callback
+    def async_register_engine(self, engine, provider, config):
+        """Register a TTS provider."""
+        provider.hass = self.hass
+        provider.use_cache = config.get(CONF_CACHE)
+        provider.language = config.get(CONF_LANG)
+        cache_dir_key = config.get(CONF_CACHE_DIR)
```

Comment:
Why would we have config related to platforms around caching, shouldn't we just always cache?